guides/source/sign_up.md

```diff
@@ -0,0 +1,1652 @@
+* How to create a nested layout
+* How to separate controllers by role
+* How to write tests for users with different roles
+
+--------------------------------------------------------------------------------
+
+Introduction
+------------
+
+One of the most common features to add to any application is a sign up process for registering new users. The e-commerce application we've built so far only has authentication and users must be created in the Rails console or a script.
+
+This feature is a required foundation for adding other features. For example, to let users create wishlists, they will need to be able to sign up first before they can create a wishlist associated with their account.
+
+Let's get started!
+
+Adding Sign Up
+--------------
+
+We've already used the Rails authentication generator to allow users to login to their accounts. The generator created a `User` model with `email_address:string` and `password_digest:string` columns in the database. It also added `has_secure_password` to the `User` model which handles passwords and confirmations. This takes care of most of what we need to add sign up to our application.
+
+### Adding Names To Users
+
+One other feature we probably want for users is to collect their name on sign up. Let's start by adding `first_name` and `last_name` columns to the database.
+
+In the terminal, create a migration with these columns:
+
+```bash
+$ rails g migration AddNamesToUsers first_name:string last_name:string
+```
+
+Then migrate the database:
+
+```bash
+$ rails db:migrate
+```
+
+We can also add validations to require these name fields on every `User`. We'll also want to display the user's full name so let's add a method to combine `first_name` and `last_name` too.
+
+Open `app/models/user.rb` and add the following:
+
+```ruby#7-11
+class User < ApplicationRecord
+  has_secure_password
+  has_many :sessions, dependent: :destroy
+
+  normalizes :email_address, with: ->(e) { e.strip.downcase }
+
+  validates :first_name, :last_name, presence: true
+
+  def full_name
+    "#{first_name} #{last_name}"
+  end
+end
+```
+
+Next, let's add sign up so we can register new users.
+
+### Sign Up Routes & Controller
+
+We now have all the necessary columns in our database to register new users. The next step is to create a route for sign up and the matching controller.
+
+In `config/routes.rb`, let's add a resource for sign up:
+
+```ruby#3
+resource :session
+resources :passwords, param: :token
+resource :sign_up
+```
+
+We're using a singular resource here because we want a singular route for `/sign_up`.
+
+This route directs requests to `app/controllers/sign_ups_controller.rb` so let's create that now.
+
+```ruby
+class SignUpsController < ApplicationController
+  def show
+    @user = User.new
+  end
+end
+```
+
+We're using the `show` action to create a new `User` instance to display the sign up form.
+
+Let's create the form next at `app/views/sign_ups/show.html.erb`:
```

Comment:
Again, do they need to create the file first or is it automatic? (Asking because I honestly forget.)